core/src/main/java/org/jruby/RubyHash.java

```diff
@@ -641,18 +641,26 @@ public boolean matches(final RubyHashEntry entry, final Object obj) {
          return newTable;
     }
 
-    public static abstract class Visitor {
+    public static abstract class VisitorWithState<T> {
+        public abstract void visit(ThreadContext context, RubyHash self, IRubyObject key, IRubyObject value, int index, T state);
+    }
+
+    public static abstract class Visitor extends VisitorWithState {
+        public void visit(ThreadContext context, RubyHash self, IRubyObject key, IRubyObject value, int index, Object state) {
+            visit(key, value);
+        }
         public abstract void visit(IRubyObject key, IRubyObject value);
     }
 
-    public void visitAll(Visitor visitor) {
```

Comment:
could we have the previous `visitAll` method (as well) for compatibility e.g. jruby-rack is using it.